src/protocols/bdx/tests/TestBdxTransferSession.cpp

```diff
@@ -0,0 +1,769 @@
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    TLV::TLVReader reader;
+    char tmp[64]        = { 0 };
+    uint32_t readLength = 0;
+    VerifyOrExit(sizeof(tmp) > len, err = CHIP_ERROR_INTERNAL);
+
+    reader.Init(dataStart, len);
+    err = reader.Next();
+
+    VerifyOrExit(reader.GetTag() == tlvListTag, err = CHIP_ERROR_INTERNAL);
+
+    // Metadata must have a top-level list
+    {
+        TLV::TLVReader listReader;
+        err = reader.OpenContainer(listReader);
+        SuccessOrExit(err);
+
+        err = listReader.Next();
+        SuccessOrExit(err);
+
+        VerifyOrExit(listReader.GetTag() == tlvStrTag, err = CHIP_ERROR_INTERNAL);
+        readLength = listReader.GetLength();
+        VerifyOrExit(readLength == expectedLen, err = CHIP_ERROR_INTERNAL);
+        err = listReader.GetString(tmp, sizeof(tmp));
+        SuccessOrExit(err);
+        VerifyOrExit(!memcmp(expected, tmp, readLength), err = CHIP_ERROR_INTERNAL);
+
+        err = reader.CloseContainer(listReader);
+        SuccessOrExit(err);
+    }
+
+exit:
+    return err;
+}
+
+// Helper method for verifying that a PacketBufferHandle contains a valid BDX header and message type matches expected.
+void VerifyBdxMessageType(nlTestSuite * inSuite, void * inContext, const System::PacketBufferHandle & msg, MessageType expected)
+{
+    CHIP_ERROR err      = CHIP_NO_ERROR;
+    uint16_t headerSize = 0;
+    PayloadHeader payloadHeader;
+
+    if (msg.IsNull())
+    {
+        NL_TEST_ASSERT(inSuite, false);
+        return;
+    }
+
+    err = payloadHeader.Decode(msg->Start(), msg->DataLength(), &headerSize);
+    NL_TEST_ASSERT(inSuite, err == CHIP_NO_ERROR);
+    NL_TEST_ASSERT(inSuite, payloadHeader.GetProtocolID() == Protocols::kProtocol_BDX);
+    NL_TEST_ASSERT(inSuite, payloadHeader.GetMessageType() == expected);
+}
+
+// Helper method for verifying that a PacketBufferHandle contains a valid StatusReport message and contains a specific StatusCode.
+void VerifyStatusReport(nlTestSuite * inSuite, void * inContext, const System::PacketBufferHandle & msg, StatusCode code)
+{
+    CHIP_ERROR err      = CHIP_NO_ERROR;
+    uint16_t headerSize = 0;
+    PayloadHeader payloadHeader;
+    uint16_t generalCode  = 0;
+    uint16_t vendorId     = 0;
+    uint16_t protocolId   = 0;
+    uint16_t protocolCode = 0;
+
+    if (msg.IsNull())
+    {
+        NL_TEST_ASSERT(inSuite, false);
+        return;
+    }
+
+    err = payloadHeader.Decode(msg->Start(), msg->DataLength(), &headerSize);
+    NL_TEST_ASSERT(inSuite, err == CHIP_NO_ERROR);
+    NL_TEST_ASSERT(inSuite, payloadHeader.GetProtocolID() == Protocols::kProtocol_Protocol_Common);
+    NL_TEST_ASSERT(inSuite, payloadHeader.GetMessageType() == static_cast<uint8_t>(Protocols::Common::MsgType::StatusReport));
+
+    Encoding::LittleEndian::Reader reader(msg->Start() + headerSize, msg->DataLength());
+    err = reader.Read16(&generalCode).Read16(&vendorId).Read16(&protocolId).Read16(&protocolCode).StatusCode();
```

Comment:
nit: if the reader is little endian, I believe a value of "(vendorid << 16) | protocolid" may actually be reversed if you want to split it out as an independent read.

Reply:
I don't think chip is splitting the concept of vendor + protocol, there is only 'protocolid' that is 32-bit and we happen to allocate 16-bit prefixes to vendors.

Reply:
Yup, BDX doesn't need to split out the vendor ID for any reason. I'll change it to just read the whole 32 bits as the profile ID.